// react-redux-grid/dist/constants/ActionTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var PAGE_LOCAL = exports.PAGE_LOCAL = '@@react-redux-grid/PAGE_LOCAL';
var PAGE_REMOTE = exports.PAGE_REMOTE = '@@react-redux-grid/PAGE_REMOTE';

var SORT_LOCAL = exports.SORT_LOCAL = '@@react-redux-grid/SORT_LOCAL';
var SORT_REMOTE = exports.SORT_REMOTE = '@@react-redux-grid/SORT_REMOTE';

var GET_DATA = exports.GET_DATA = '@@react-redux-grid/GET_DATA';
var SET_DATA = exports.SET_DATA = '@@react-redux-grid/SET_DATA';
var SORT_DATA = exports.SORT_DATA = '@@react-redux-grid/SORT_DATA';
var INSERT_ROW = exports.INSERT_ROW = '@@react-redux-grid/INSERT_ROW';

var SET_SELECTION = exports.SET_SELECTION = '@@react-redux-grid/SET_SELECTION';
var SELECT_ALL = exports.SELECT_ALL = '@@react-redux-grid/SELECT_ALL';
var DESELECT_ALL = exports.DESELECT_ALL = '@@react-redux-grid/DESELECT_ALL';
var SELECT_ROW = exports.SELECT_ROW = '@@react-redux-grid/SELECT_ROW';
var DESELECT_ROW = exports.DESELECT_ROW = '@@react-redux-grid/DESELECT_ROW';

var SET_COLUMNS = exports.SET_COLUMNS = '@@react-redux-grid/SET_COLUMNS';
var SET_SORT_DIRECTION = exports.SET_SORT_DIRECTION = '@@react-redux-grid/SET_SORT_DIRECTION';

var SET_LOADING_STATE = exports.SET_LOADING_STATE = '@@react-redux-grid/SET_LOADING_STATE';

var DISMISS_ERROR = exports.DISMISS_ERROR = '@@react-redux-grid/DISMISS_ERROR';
var ERROR_OCCURRED = exports.ERROR_OCCURRED = '@@react-redux-grid/ERROR_OCCURRED';

var SHOW_MENU = exports.SHOW_MENU = '@@react-redux-grid/SHOW_MENU';
var HIDE_MENU = exports.HIDE_MENU = '@@react-redux-grid/HIDE_MENU';

var RESIZE_COLUMNS = exports.RESIZE_COLUMNS = '@@react-redux-grid/RESIZE_COLUMNS';
var REORDER_COLUMN = exports.REORDER_COLUMN = '@@react-redux-grid/REORDER_COLUMN';
var HIDE_HEADER = exports.HIDE_HEADER = '@@react-redux-grid/HIDE_HEADER';

var NO_EVENT = exports.NO_EVENT = '@@react-redux-grid/NO_EVENT';

var EDIT_ROW = exports.EDIT_ROW = '@@react-redux-grid/EDIT_ROW';
var REPOSITION_EDITOR = exports.REPOSITION_EDITOR = '@@react-redux-grid/REPOSITION_EDITOR';
var REMOVE_ROW = exports.REMOVE_ROW = '@@react-redux-grid/REMOVE_ROW';
var ADD_NEW_ROW = exports.ADD_NEW_ROW = '@@react-redux-grid/ADD_NEW_ROW';
var CANCEL_ROW = exports.CANCEL_ROW = '@@react-redux-grid/CANCEL_ROW';
var SAVE_ROW = exports.SAVE_ROW = '@@react-redux-grid/SAVE_ROW';
var UPDATE_ROW = exports.UPDATE_ROW = '@@react-redux-grid/UPDATE_ROW';
var ROW_VALUE_CHANGE = exports.ROW_VALUE_CHANGE = '@@react-redux-grid/ROW_VALUE_CHANGE';
var DISMISS_EDITOR = exports.DISMISS_EDITOR = '@@react-redux-grid/DISMISS_EDITOR';

var REMOVE_TOOLBAR = exports.REMOVE_TOOLBAR = '@@react-redux-grid/REMOVE_TOOLBAR';

var SET_FILTER_VALUE = exports.SET_FILTER_VALUE = '@@react-redux-grid/SET_FILTER_VALUE';
var FILTER_DATA = exports.FILTER_DATA = '@@react-redux-grid/FILTER_DATA';
var CLEAR_FILTER_LOCAL = exports.CLEAR_FILTER_LOCAL = '@@react-redux-grid/CLEAR_FILTER_LOCAL';
var CLEAR_FILTER_REMOTE = exports.CLEAR_FILTER_REMOTE = '@@react-redux-grid/CLEAR_FILTER_REMOTE';
var SHOW_FILTER_MENU = exports.SHOW_FILTER_MENU = '@@react-redux-grid/SHOW_FILTER_MENU';
var SET_FILTER_MENU_VALUES = exports.SET_FILTER_MENU_VALUES = '@@react-redux-grid/SET_FILTER_MENU_VALUES';

var MOVE_NODE = exports.MOVE_NODE = '@@react-redux-grid/MOVE_NODE';

var SET_TREE_NODE_VISIBILITY = exports.SET_TREE_NODE_VISIBILITY = '@@react-redux-grid/SET_TREE_NODE_VISIBILITY';

var SET_TREE_DATA_PARTIAL = exports.SET_TREE_DATA_PARTIAL = '@@react-redux-grid/SET_TREE_DATA_PARTIAL';
;

var _temp = function () {
    if (typeof __REACT_HOT_LOADER__ === 'undefined') {
        return;
    }

    __REACT_HOT_LOADER__.register(PAGE_LOCAL, 'PAGE_LOCAL', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(PAGE_REMOTE, 'PAGE_REMOTE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SORT_LOCAL, 'SORT_LOCAL', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SORT_REMOTE, 'SORT_REMOTE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(GET_DATA, 'GET_DATA', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_DATA, 'SET_DATA', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SORT_DATA, 'SORT_DATA', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(INSERT_ROW, 'INSERT_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_SELECTION, 'SET_SELECTION', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SELECT_ALL, 'SELECT_ALL', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(DESELECT_ALL, 'DESELECT_ALL', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SELECT_ROW, 'SELECT_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(DESELECT_ROW, 'DESELECT_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_COLUMNS, 'SET_COLUMNS', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_SORT_DIRECTION, 'SET_SORT_DIRECTION', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_LOADING_STATE, 'SET_LOADING_STATE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(DISMISS_ERROR, 'DISMISS_ERROR', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(ERROR_OCCURRED, 'ERROR_OCCURRED', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SHOW_MENU, 'SHOW_MENU', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(HIDE_MENU, 'HIDE_MENU', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(RESIZE_COLUMNS, 'RESIZE_COLUMNS', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(REORDER_COLUMN, 'REORDER_COLUMN', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(HIDE_HEADER, 'HIDE_HEADER', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(NO_EVENT, 'NO_EVENT', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(EDIT_ROW, 'EDIT_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(REPOSITION_EDITOR, 'REPOSITION_EDITOR', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(REMOVE_ROW, 'REMOVE_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(ADD_NEW_ROW, 'ADD_NEW_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(CANCEL_ROW, 'CANCEL_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SAVE_ROW, 'SAVE_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(UPDATE_ROW, 'UPDATE_ROW', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(ROW_VALUE_CHANGE, 'ROW_VALUE_CHANGE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(DISMISS_EDITOR, 'DISMISS_EDITOR', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(REMOVE_TOOLBAR, 'REMOVE_TOOLBAR', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_FILTER_VALUE, 'SET_FILTER_VALUE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(FILTER_DATA, 'FILTER_DATA', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(CLEAR_FILTER_LOCAL, 'CLEAR_FILTER_LOCAL', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(CLEAR_FILTER_REMOTE, 'CLEAR_FILTER_REMOTE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SHOW_FILTER_MENU, 'SHOW_FILTER_MENU', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_FILTER_MENU_VALUES, 'SET_FILTER_MENU_VALUES', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(MOVE_NODE, 'MOVE_NODE', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_TREE_NODE_VISIBILITY, 'SET_TREE_NODE_VISIBILITY', 'src/constants/ActionTypes.js');

    __REACT_HOT_LOADER__.register(SET_TREE_DATA_PARTIAL, 'SET_TREE_DATA_PARTIAL', 'src/constants/ActionTypes.js');
}();

;
