// react-redux-grid/dist/index.js
'use strict';

var _Grid = require('./components/Grid');

var _Grid2 = _interopRequireDefault(_Grid);

var _store = require('./store/store');

var _store2 = _interopRequireDefault(_store);

var _reducers = require('./reducers');

var _actions = require('./actions');

var _GridConstants = require('./constants/GridConstants');

var _ActionTypes = require('./constants/ActionTypes');

var ActionTypes = _interopRequireWildcard(_ActionTypes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var modules = {
    Actions: _actions.Actions,
    Grid: _Grid2.default,
    Reducers: _reducers.Reducers,
    applyGridConfig: _GridConstants.applyGridConfig,
    ActionTypes: ActionTypes,
    Store: _store2.default
};

module.exports = modules;
;

var _temp = function () {
    if (typeof __REACT_HOT_LOADER__ === 'undefined') {
        return;
    }

    __REACT_HOT_LOADER__.register(modules, 'modules', 'src/index.js');
}();

;
